### PR TITLE
Set filter to preloading data source for partition pruning

### LINKED ISSUE #7381: Add dynamic filter for preloading data source
### Description

`pendingDynamicFilters_` is evaluated to skip partitions in `HiveDataSource::addSplit` --> `SplitReader::prepareSplit` --> `testFilters`. They are added in `dataSource_` but not preloading data sources.
This is the case we observed in TPCDS query59, which impacting the raw input bytes after enabling split preloading.

Comparing:
image
Add the filters:
image

### PR DESCRIPTION
**Pull request by zhli1142015**
Fixes #7381

### CODE REVIEW

File: velox/exec/TableScan.cpp [hunk 1]
```diff
@@ -92,6 +92,7 @@ RowVectorPtr TableScan::getOutput() {
 
       if (!split.hasConnectorSplit()) {
         noMoreSplits_ = true;
+        pendingDynamicFilters_.clear();
```
Comment (Yuhta) [inline velox/exec/TableScan.cpp:95]: I think it's still more clear that we clear the pending filters after passing them to data source

File: velox/exec/TableScan.cpp [hunk 2]
```diff
@@ -256,6 +256,7 @@ void TableScan::preload(std::shared_ptr<connector::ConnectorSplit> split) {
        ctx = operatorCtx_->createConnectorQueryCtx(
            split->connectorId, planNodeId(), connectorPool_),
        task = operatorCtx_->task(),
+       pendingDynamicFilters = pendingDynamicFilters_,
```
Comment (Yuhta) [inline velox/exec/TableScan.cpp:259]: Just do `std::move(pendingDynamicFilters = pendingDynamicFilters_)`

**Review by Yuhta — COMMENTED** [review]
Thanks for the fix, is it possible to add a unit test?

**Comment by zhli1142015** [comment]
> Thanks for the fix, is it possible to add a unit test?

Add a UT to assert the split skipping happens in preloaded splits. Thanks.

File: velox/exec/TableScan.cpp (hunk 2 above)
Comment (zhli1142015) [reply velox/exec/TableScan.cpp]: I think you mean we should move not copy the `pendingDynamicFilters_`. maybe we should use by-reference capture here.

File: velox/exec/TableScan.cpp (hunk 1 above)
Comment (zhli1142015) [reply velox/exec/TableScan.cpp:95]: `pendingDynamicFilters_` is used in `preload` to intitalize different data source object for preloading splits, I think it should be clear after no more splits. Thanks.

**Comment by zhli1142015** [comment]
Hello @Yuhta , could you please check again?

File: velox/exec/TableScan.cpp (hunk 1 above)
Comment (Yuhta) [reply velox/exec/TableScan.cpp:95]: Emm I see what you want to do.  But merging filters on every split could cause performance regression. I think we need to do it a little differently.  Maybe when we do the preload, we should pass in the scan spec from the current `dataSource_` to use in preloading data source as well.

Comment (zhli1142015) [reply velox/exec/TableScan.cpp]: Hello @Yuhta ,
I tried this proposal, the `scanSpec_` can't be used by multi threads same time, even only the `filter_` field in `ScanSpec`. 
For your concern about merging filters on every split, I add two APIs in `DataSource` to get and set the filter for the specific `outputChannel`, then we only need to set the merged filters to the preloading data sources. This should avoid this problem.
Please help to chekc again, thanks.

File: velox/connectors/hive/HiveDataSource.h [hunk 3]
```diff
@@ -71,6 +71,19 @@ class HiveDataSource : public DataSource {
 
   int64_t estimatedRowSize() override;
 
+  std::unique_ptr<common::Filter> getFilter(
+      column_index_t outputChannel) override {
+    auto& fieldSpec = scanSpec_->getChildByChannel(outputChannel);
+    return fieldSpec.filter()->clone();
```
Comment (Yuhta) [inline velox/connectors/hive/HiveDataSource.h:77]: The filter should be MT safe as far as I see, all the methods should be read only.  Can you try using a `shared_ptr<const Filter>` to avoid deep copy?

File: velox/exec/tests/HashJoinTest.cpp [hunk 4]
```diff
@@ -4658,6 +4659,87 @@ TEST_F(HashJoinTest, dynamicFiltersWithSkippedSplits) {
...
+  auto outputType =
+      ROW({"n1_0", "n1_1", "n1_2"}, {BIGINT(), BIGINT(), BIGINT()});
+  std::shared_ptr<connector::hive::HiveTableHandle> tableHandle =
+      makeTableHandle();
+  ColumnHandleMap assignments = {
+      {"n1_0", regularColumn("c0", BIGINT())},
+      {"n1_1", regularColumn("c1", BIGINT())},
+      {"n1_2", partitionKey("k", BIGINT())}};
+
+  FLAGS_split_preload_per_driver = 2;
```
[the remark this answers is not in the document]
Comment (zhli1142015) [reply velox/exec/tests/HashJoinTest.cpp:4710]: Updated, thanks.

File: velox/exec/TableScan.cpp [hunk 5]
```diff
@@ -256,6 +259,7 @@ void TableScan::preload(std::shared_ptr<connector::ConnectorSplit> split) {
        ctx = operatorCtx_->createConnectorQueryCtx(
            split->connectorId, planNodeId(), connectorPool_),
        task = operatorCtx_->task(),
+       &mergedDynamicFilters = mergedDynamicFilters_,
```
[the remark this answers is not in the document]
Comment (zhli1142015) [reply velox/exec/TableScan.cpp]: Changed, thanks.

File: velox/connectors/hive/HiveDataSource.h (hunk 3 above)
Comment (zhli1142015) [reply velox/connectors/hive/HiveDataSource.h]: Yes, you are right, changed.

**Comment by zhli1142015** [comment]
Hello @Yuhta ,
I addressed all comments, could you please continue to take a look?
Thanks.

**Comment by zhli1142015** [comment]
The fuzzer failure looks realted to https://github.com/facebookincubator/velox/issues/7856

```
I20231206 04:56:20.786527 29572 ExpressionVerifier.cpp:97] Executing expression 0 : try(translate(regex_replace("c0",md5("c1"),Y`%2&fl*ZC`d`"F@-j/T2),"c0",^>XIvcHADU#jT[eE&6[JR1gy$vj!4W%9~RUE/W`>Jq,hRy4xUWdOSg-jI))
E20231206 04:56:20.811030 29572 Exceptions.h:69] Line: ../../velox/expression/tests/ExpressionVerifier.cpp:77, Function:operator(), Expression: left->equalValueAt(right.get(), row, row) Different results at idx '0': '[0->71] null' vs. '^>XI>cHADU#jT[eE&6[JJ1gy$vA!y#%9#RUEyg`>JARhgy4jU[dOeg-jj-E`EjJ&1J19`', Source: RUNTIME, ErrorCode: INVALID_STATE
```

File: velox/connectors/Connector.h [hunk 6]
```diff
@@ -226,6 +226,17 @@ class DataSource {
   virtual int64_t estimatedRowSize() {
     return kUnknownRowSize;
   }
+
+  virtual std::shared_ptr<common::Filter> getFilter(
```
Comment (mbasmanova) [inline velox/connectors/Connector.h:230]: This PR changes the Connector API. Please, update PR description to describe the changes and add comments for the newly introduced APIs.

File: velox/exec/tests/HashJoinTest.cpp [hunk 7]
```diff
@@ -4670,6 +4671,88 @@ TEST_F(HashJoinTest, dynamicFiltersWithSkippedSplits) {
...
+        SCOPED_TRACE(fmt::format("hasSpill:{}", hasSpill));
+        if (!hasSpill) {
+          ASSERT_EQ(1, getFiltersProduced(task, 1).sum);
+          ASSERT_EQ(1, getFiltersAccepted(task, 0).sum);
+          ASSERT_EQ(4, getOperatorRuntimeStats(task, 0, "skippedSplits").sum);
+          ASSERT_LT(0, getOperatorRuntimeStats(task, 0, "preloadedSplits").sum);
+        }
+      })
+      .run();
+  FLAGS_split_preload_per_driver = oldSplitPreload;
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:4753]: This is an anti-pattern. Use RAII.

File: velox/connectors/Connector.h (hunk 6 above)
Comment (zhli1142015) [reply velox/connectors/Connector.h]: Updated, thanks.

File: velox/connectors/Connector.h [hunk 8]
```diff
@@ -226,6 +226,21 @@ class DataSource {
...
+  // Returns the filter for the column identified by @param outputChannel.
+  virtual std::shared_ptr<common::Filter> getColumnFilter(
+      column_index_t outputChannel) {
+    VELOX_UNSUPPORTED("getFilter");
+  }
+
+  // Sets the filter to the column identified by @param outputChannel.
+  // This can be called when passing the column filters to a preloading data
+  // source. The purpose of this is to avoid the overhead of filters merging.
+  virtual void setColumnFilter(
```
Comment (mbasmanova) [inline velox/connectors/Connector.h:239]: How is this different from addDynamicFilter? Having these two APIs is confusing. Please, update PR description to provide overall context for this change.

File: velox/exec/tests/utils/FlagUpdater.h [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+namespace facebook::velox::exec::test {
+
+// Generic RAII class to update GFlag.
+template <typename T>
+class FlagUpdater {
```
Comment (mbasmanova) [inline velox/exec/tests/utils/FlagUpdater.h:21]: Maybe ScopedGFlag?

File: velox/exec/tests/HashJoinTest.cpp [hunk 10]
```diff
@@ -4670,6 +4672,86 @@ TEST_F(HashJoinTest, dynamicFiltersWithSkippedSplits) {
   }
 }
 
+TEST_F(HashJoinTest, dynamicFiltersAppliedToPreloadedSplits) {
+  vector_size_t size = 1000;
+  const int32_t numSplits = 5;
+  std::vector<RowVectorPtr> probeVectors;
+  probeVectors.reserve(numSplits);
+  FlagUpdater(FLAGS_split_preload_per_driver, 3);
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:4680]: Are these other tests that use split_preload_per_driver flag and therefore need to be updated to use the new RAII class? If so, would you extract the RAII into a separate PR and updating existing tests?

Comment (mbasmanova) [reply velox/exec/tests/HashJoinTest.cpp:4680]: Got it. Let's make a new PR to replace this gflag with a configuration property. Then come back to this PR.

File: velox/connectors/Connector.h (hunk 8 above)
Comment (mbasmanova) [reply velox/connectors/Connector.h:239]: @zhli1142015 Thank you for this additional context. Is there a GitHub issue about this? Sounds like we need to iterate on the design a bit and it will be easier to do in a GitHub issue.

File: velox/exec/tests/HashJoinTest.cpp (hunk 10 above)
Comment (zhli1142015) [reply velox/exec/tests/HashJoinTest.cpp:4680]: https://github.com/facebookincubator/velox/pull/7931, thanks.

**Comment by zhli1142015** [comment]
The failure should be not related to this change, thanks.
```
I20231211 03:18:16.575408 29965 AggregationFuzzer.cpp:870] ==============================> Started iteration 30 (seed: 3731594752)
terminate called after throwing an instance of 'facebook::velox::dwio::common::exception::LoggedException'
  what():  Exception: VeloxException
Error Source: RUNTIME
Error Code: UNKNOWN
Reason: not supported yet HUGEINT
Retriable: False
Function: create
File: ../../velox/dwio/dwrf/writer/ColumnWriter.cpp
Line: 2082
Stack trace:
# 0  _ZN8facebook5velox7process10StackTraceC1Ei
# 1  _ZN8facebook5velox14VeloxException5State4makeIZNS1_C4EPKcmS5_St17basic_string_viewIcSt11char_traitsIcEES9_S9_S9_bNS1_4TypeES9_EUlRT_E_EESt10shared_ptrIKS2_ESA_SB_
# 2  _ZN8facebook5velox14VeloxExceptionC1EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bNS1_4TypeES7_
# 3  _ZN8facebook5velox4dwio6common9exception15LoggedExceptionC1EPKcmS6_S6_RKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESE_SE_
# 4  _ZN8facebook5velox4dwrf16BaseColumnWriter6createERNS1_13WriterContextERKNS0_4dwio6common10TypeWithIdEjSt8functionIFvRNS1_12IndexBuilderEEE
# 5  _ZN8facebook5velox4dwrf16BaseColumnWriter6createERNS1_13WriterContextERKNS0_4dwio6common10TypeWithIdEjSt8functionIFvRNS1_12IndexBuilderEEE
# 6  _ZN8facebook5velox4dwrf6WriterC2ESt10unique_ptrINS0_4dwio6common8FileSinkESt14default_deleteIS6_EERKNS1_13WriterOptionsESt10shared_ptrINS0_6memory10MemoryPoolEE
# 7  _ZN8facebook5velox4dwrf6WriterC1ESt10unique_ptrINS0_4dwio6common8FileSinkESt14default_deleteIS6_EERKNS1_13WriterOptionsE
# 8  _ZN8facebook5velox4exec4test12_GLOBAL__N_112_GLOBAL__N_111writeToFileERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERKSt10shared_ptrINS0_10BaseVectorEEPNS0_6memory10MemoryPoolE
# 9  _ZN8facebook5velox4exec4test12_GLOBAL__N_117AggregationFuzzer10makeSplitsERKSt6vectorISt10shared_ptrINS0_9RowVectorEESaIS8_EERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
# 10 _ZN8facebook5velox4exec4test12_GLOBAL__N_117AggregationFuzzer17verifyAggregationERKSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaISB_EESF_SF_RKS5_ISt10shared_ptrINS0_9RowVectorEESaISI_EEbRKS5_ISG_INS2_14ResultVerifierEESaISO_EE
# 11 _ZN8facebook5velox4exec4test12_GLOBAL__N_117AggregationFuzzer2goEv
# 12 _ZN8facebook5velox4exec4test15aggregateFuzzerESt13unordered_mapINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt6vectorISt10shared_ptrINS1_26AggregateFunctionSignatureEESaISD_EESt4hashIS9_ESt8equal_toIS9_ESaISt4pairIKS9_SF_EEEmRKS3_IS9_SB_INS2_14ResultVerifierEESH_SJ_SaISK_ISL_SQ_EEERKS3_IS9_SB_INS2_14InputGeneratorEESH_SJ_SaISK_ISL_SX_EEENS0_12VectorFuzzer7Options18TimestampPrecisionERKS3_IS9_S9_SH_SJ_SaISK_ISL_S9_EEERKSt8optionalIS9_ESt10unique_ptrINS2_20ReferenceQueryRunnerESt14default_deleteIS1G_EE
# 13 _ZN8facebook5velox4exec4test23AggregationFuzzerRunner9runFuzzerEmRKSt8optionalINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEESt10unique_ptrINS2_20ReferenceQueryRunnerESt14default_deleteISF_EERKNS3_7OptionsE
# 14 _ZN8facebook5velox4exec4test23AggregationFuzzerRunner3runEmSt10unique_ptrINS2_20ReferenceQueryRunnerESt14default_deleteIS5_EERKNS3_7OptionsE
# 15 main
# 16 __libc_start_main
# 17 _start
```

**Comment by zhli1142015** [comment]
@mbasmanova and @Yuhta , could you please help to review this PR? I made the change according to the comment: https://github.com/facebookincubator/velox/issues/7381#issuecomment-1845962921

File: velox/dwio/common/ScanSpec.h [hunk 11]
```diff
@@ -280,7 +280,20 @@ class ScanSpec {
...
+  const std::shared_ptr<common::Filter>& getFilterByChannel(
+      column_index_t channel) const {
+    return getChildByChannel(channel).filter_;
+  }
+
+  void addChildFilterByChannel(column_index_t channel, common::Filter& filter) {
+    getChildByChannel(channel).addFilter(filter);
+  }
+
+  void setChildFilterByChannel(column_index_t channel, common::Filter& filter) {
```
Comment (Yuhta) [inline velox/dwio/common/ScanSpec.h:294]: Take `shared_ptr` to avoid deep copy

File: velox/connectors/hive/HiveDataSource.h [hunk 12]
```diff
@@ -71,6 +71,11 @@ class HiveDataSource : public DataSource {
 
   int64_t estimatedRowSize() override;
 
+  std::shared_ptr<common::Filter> getColumnFilter(
```
Comment (Yuhta) [inline velox/connectors/hive/HiveDataSource.h:74]: `const std::shared_ptr<common::Filter>&`

Comment (zhli1142015) [reply velox/connectors/hive/HiveDataSource.h]: Updated, thanks.

File: velox/dwio/common/ScanSpec.h (hunk 11 above)
Comment (zhli1142015) [reply velox/dwio/common/ScanSpec.h]: Updated, thanks.

File: velox/connectors/hive/HiveDataSource.cpp [hunk 13]
```diff
@@ -702,6 +704,8 @@ void HiveDataSource::setFromDataSource(
 
   split_ = std::move(source->split_);
   if (source->splitReader_ && source->splitReader_->emptySplit()) {
+    runtimeStats_.skippedSplits += source->runtimeStats_.skippedSplits;
```
Comment (Yuhta) [inline velox/connectors/hive/HiveDataSource.cpp:707]: This will be double counting, no?  It already added in `SplitReader::prepareSplit`

Comment (zhli1142015) [reply velox/connectors/hive/HiveDataSource.cpp:707]: These metrics are collected within the `prepareSplit` function.
The purpose of this section of code is to pass the metrics obtained during the 'preloading' data source (captured by `runtimeStats_`) to the main data source (`dataSource_`).
This is found by new added UT. Without this line, the result of `prepareSplit` in new UT is consistently 0, whereas the expected result is 4. 
The correction ensures that the runtime metrics are properly transferred to `dataSource_`, leading to the desired outcome.

Thanks.

**Review by mbasmanova — CHANGES_REQUESTED** [review]
@zhli1142015 This PR is a bit confusing and hard to understand especially because there is no description. My impression is that the main trouble is that you want to avoid the cost of combining static and dynamic filters for each split. Do you have some evidence that explains why this is so important? Please, update the PR description to provide all the necessary context and add comments in code to explain why a simpler more intuitive solution of just calling existing addDynamicFilter for preloading sources is not good enough.

I made some suggestions on API changes based on my current understanding below.

**Comment by zhli1142015** [comment]
> a bit confusing and hard to understand especially because there is no description. My impression is that the main trouble is that you want to avoid the cost of combining static and dynamic filters for each split. Do you have some evidence that explains why this is so important? Please, update the PR description to provide all the necessary context and add comments in code to explain why a simpler more intuitive solution of just calling existing addDynamicFilter for preloading sources is not good enough.

Hello @mbasmanova ,
I have no evidence to prove that this is an issue. On the contrary, within our internal branch, we use the existing addDynamicFilter, and in our testing, we haven't observed any problems. The reason for doing this is because of this comment  https://github.com/facebookincubator/velox/pull/7382#discussion_r1390006231. I assumed this approach was taken due to some presto limitation not spark side. If we can call existing addDynamicFilter for preloading sources, then change should be simple and understandable enough.

**Comment by mbasmanova** [comment]
@zhli1142015 

> I have no evidence to prove that this is an issue. On the contrary, within our internal repo, we use the existing addDynamicFilter, and in our testing, we haven't observed any problems. The reason for doing this is because of this comment https://github.com/facebookincubator/velox/pull/7382#discussion_r1390006231. I assumed this approach was taken due to some presto limitation not spark side. If we can call existing addDynamicFilter for preloading sources, then [change](https://github.com/facebookincubator/velox/pull/7382/commits/c7665ae1a35143a24d3bba2ac3c222f38885c293) should be simple and understandable enough.

Thank you for clarifying.

> But merging filters on every split could cause performance regression. 

@Yuhta Jimmy, looks like this may not be an issue. Trying to optimize for this hypothetical problem leads to pretty convoluted code. Perhaps, we don't need that can use a simpler alternative instead.

**Comment by zhli1142015** [comment]
> @zhli1142015
> 
> > I have no evidence to prove that this is an issue. On the contrary, within our internal repo, we use the existing addDynamicFilter, and in our testing, we haven't observed any problems. The reason for doing this is because of this comment [#7382 (comment)](https://github.com/facebookincubator/velox/pull/7382#discussion_r1390006231). I assumed this approach was taken due to some presto limitation not spark side. If we can call existing addDynamicFilter for preloading sources, then [change](https://github.com/facebookincubator/velox/pull/7382/commits/c7665ae1a35143a24d3bba2ac3c222f38885c293) should be simple and understandable enough.
> 
> Thank you for clarifying.
> 
> > But merging filters on every split could cause performance regression.
> 
> @Yuhta Jimmy, looks like this may not be an issue. Trying to optimize for this hypothetical problem leads to pretty convoluted code. Perhaps, we don't need that can use a simpler alternative instead.

Hello @mbasmanova ,
I change this PR back to utilize existing addDynamicFilter fucntion, could you please help check?
Thanks.

File: velox/exec/TableScan.cpp [hunk 14]
```diff
@@ -256,6 +256,7 @@ void TableScan::preload(std::shared_ptr<connector::ConnectorSplit> split) {
        ctx = operatorCtx_->createConnectorQueryCtx(
            split->connectorId, planNodeId(), connectorPool_),
        task = operatorCtx_->task(),
+       &pendingDynamicFilters = pendingDynamicFilters_,
```
Comment (mbasmanova) [inline velox/exec/TableScan.cpp:259]: Do we need a reference here? We pass everything else by value.

File: velox/exec/tests/HashJoinTest.cpp [hunk 15]
```diff
@@ -4669,6 +4670,87 @@ TEST_F(HashJoinTest, dynamicFiltersWithSkippedSplits) {
...
+      splits.emplace(nodeId, probeSplits);
+      return splits;
+    };
+  };
+  auto outputType =
+      ROW({"n1_0", "n1_1", "n1_2"}, {BIGINT(), BIGINT(), BIGINT()});
+  ColumnHandleMap assignments = {
+      {"n1_0", regularColumn("c0", BIGINT())},
+      {"n1_1", regularColumn("c1", BIGINT())},
+      {"n1_2", partitionKey("k", BIGINT())}};
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:4717]: Is it important to use partition keys columns in this test? Would you explain why?

File: velox/exec/tests/HashJoinTest.cpp [hunk 16]
```diff
@@ -4669,6 +4670,87 @@ TEST_F(HashJoinTest, dynamicFiltersWithSkippedSplits) {
...
+  HashJoinBuilder(*pool_, duckDbQueryRunner_, driverExecutor_.get())
+      .planNode(std::move(op))
+      .config(core::QueryConfig::kMaxSplitPreloadPerDriver, "3")
+      .makeInputSplits(makeInputSplits(probeScanId))
+      .referenceQuery("select t.c0 from t, u where t.c0 = u.c1")
+      .checkSpillStats(false)
+      .verifier([&](const std::shared_ptr<Task>& task, bool hasSpill) {
+        SCOPED_TRACE(fmt::format("hasSpill:{}", hasSpill));
+        if (!hasSpill) {
+          ASSERT_EQ(1, getFiltersProduced(task, 1).sum);
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:4745]: It is error prone to use hard-coded operator indices here. Can we use plan node IDs instead? We have probeScanId and we can also capture join node ID. We can then get stats via toPlanStats.

```
std::unordered_map<core::PlanNodeId, PlanNodeStats> toPlanStats(
    const TaskStats& taskStats);
```

File: velox/exec/tests/HashJoinTest.cpp [hunk 17]
```diff
@@ -4669,6 +4670,87 @@ TEST_F(HashJoinTest, dynamicFiltersWithSkippedSplits) {
...
+                         .build();
+        probeSplits.push_back(exec::Split(split));
+      }
+      SplitInput splits;
+      splits.emplace(nodeId, probeSplits);
+      return splits;
+    };
+  };
+  auto outputType =
+      ROW({"n1_0", "n1_1", "n1_2"}, {BIGINT(), BIGINT(), BIGINT()});
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:4713]: These column names are hard to read. By convention, we use t0, t1, ... and u0, u1, ..... You can specify column names as first argument of makeRowVector.

**Review by mbasmanova — COMMENTED** [review]
Looks good % some questions about the test.

File: velox/exec/tests/HashJoinTest.cpp (hunk 16 above)
Comment (zhli1142015) [reply velox/exec/tests/HashJoinTest.cpp]: Updated.

File: velox/exec/TableScan.cpp (hunk 14 above)
Comment (zhli1142015) [reply velox/exec/TableScan.cpp]: Thanks for catching this, we should pass by value here

File: velox/exec/tests/HashJoinTest.cpp (hunk 15 above)
Comment (zhli1142015) [reply velox/exec/tests/HashJoinTest.cpp]: Yes, we expect dynamic filters on the partition column to be generated. This way, we can verify whether the dynamic filter is passed to the preloaded data source by checking if the split is skipped during processing.

File: velox/exec/tests/HashJoinTest.cpp [hunk 18]
```diff
@@ -4669,6 +4670,102 @@ TEST_F(HashJoinTest, dynamicFiltersWithSkippedSplits) {
...
+    tempFiles.push_back(TempFilePath::create());
+    writeToFile(tempFiles.back()->path, rowVector);
+  }
+
+  auto makeInputSplits = [&](const core::PlanNodeId& nodeId) {
+    return [&] {
+      std::vector<exec::Split> probeSplits;
+      for (int32_t i = 0; i < numSplits; ++i) {
+        auto value = std::to_string(i);
+        auto split = facebook::velox::exec::test::HiveConnectorSplitBuilder(
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:4700]: drop facebook::velox::exec::test::

File: velox/exec/tests/HashJoinTest.cpp [hunk 19]
```diff
@@ -4669,6 +4670,102 @@ TEST_F(HashJoinTest, dynamicFiltersWithSkippedSplits) {
...
+              "",
+              {"p0"},
+              core::JoinType::kInner)
+          .capturePlanNodeId(joinNodeId)
+          .project({"p0"})
+          .planNode();
+  HashJoinBuilder(*pool_, duckDbQueryRunner_, driverExecutor_.get())
+      .planNode(std::move(op))
+      .config(core::QueryConfig::kMaxSplitPreloadPerDriver, "3")
+      .makeInputSplits(makeInputSplits(probeScanId))
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:4749]: this code is confusing as makeInputSplits is repeated twice; would you use .inputSplits(probeSplits) instead and remove makeInputSplits lambda from this test?

**Review by mbasmanova — COMMENTED** [review]
Looks good % some minor comments.

File: velox/exec/tests/HashJoinTest.cpp (hunk 19 above)
Comment (zhli1142015) [reply velox/exec/tests/HashJoinTest.cpp]: Updated, thanks.

File: velox/exec/tests/HashJoinTest.cpp (hunk 18 above)
Comment (zhli1142015) [reply velox/exec/tests/HashJoinTest.cpp]: Updated.

File: velox/exec/tests/HashJoinTest.cpp [hunk 20]
```diff
@@ -4669,6 +4670,90 @@ TEST_F(HashJoinTest, dynamicFiltersWithSkippedSplits) {
...
+
+  // Prepare build side table.
+  std::vector<RowVectorPtr> buildVectors{
+      makeRowVector({"b0"}, {makeFlatVector<int64_t>({0, numSplits})})};
+  createDuckDbTable("b", buildVectors);
+
+  // Executing the join with p1=b0, we expect a dynamic filter for p1 to prune
+  // the entire file/split. There are total of five splits, and all except the
+  // first one are expected to be pruned. The result 'preloadedSplits' > 1
+  // confirms the successful push of dynamic filters to the preloading data
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:4714]: Nice comment.

File: velox/exec/tests/HashJoinTest.cpp [hunk 21]
```diff
@@ -4669,6 +4670,90 @@ TEST_F(HashJoinTest, dynamicFiltersWithSkippedSplits) {
...
+  // source.
+  core::PlanNodeId probeScanId, joinNodeId;
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:4716]: define each variable separately

File: velox/exec/tests/HashJoinTest.cpp [hunk 22]
```diff
@@ -4669,6 +4670,90 @@ TEST_F(HashJoinTest, dynamicFiltersWithSkippedSplits) {
...
+              {"p1"},
+              {"b0"},
+              PlanBuilder(planNodeIdGenerator).values(buildVectors).planNode(),
+              "",
+              {"p0"},
+              core::JoinType::kInner)
+          .capturePlanNodeId(joinNodeId)
+          .project({"p0"})
+          .planNode();
+  SplitInput splits = {{probeScanId, probeSplits}};
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:4735]: no need for this variable

`.splits({{probeScanId, probeSplits}})`

**Review by mbasmanova — APPROVED** [review]
Thanks.

File: velox/exec/tests/HashJoinTest.cpp (hunk 21 above)
Comment (zhli1142015) [reply velox/exec/tests/HashJoinTest.cpp]: Updated, thanks.

**Comment by zhli1142015** [comment]
Hello @mbasmanova ,
Looks CI failed in internal tests, please let me know if any fix needed, thanks.

**Comment by mbasmanova** [comment]
@zhli1142015 Would you rebase?

**Comment by zhli1142015** [comment]
> @zhli1142015 Would you rebase?

Done, thanks.
